onnxscript/ir/passes/common/common_subexpression_elimination.py

```diff
@@ -0,0 +1,163 @@
+        modified = False
+        # 1. Initialize a new graph. It will be used to store the new nodes.
+        #    and replace the old graph.
+        old_graph = model.graph
+        # Values and nodes point to the old graph, so they need to be
+        # created in the new graph.
+        new_graph = ir.Graph(
+            inputs=[],
+            outputs=[],
+            nodes=[],
+            initializers=[],
+            name=old_graph.name,
+            metadata_props=old_graph.metadata_props,
+            opset_imports=old_graph.opset_imports,
+            doc_string=old_graph.doc_string,
+        )
+        # 2. Create a mapping from olds to news.
+        old_node_hash_to_new_node: dict[int, ir.Node] = {}
+        old_value_to_new_value: dict[ir.Value, ir.Value] = {}
+
+        # 3. Create inputs and initializers in the new graph
+        #    from the old graph.
+        for input in old_graph.inputs:
+            new_input = _copy_value(input)
+            new_graph.inputs.append(new_input)
+            old_value_to_new_value[input] = new_input
+            logger.debug("Adding input %s", new_input.name)
+        for initializer in old_graph.initializers.values():
+            new_initializer = _copy_value(initializer)
+            new_graph.register_initializer(new_initializer)
+            old_value_to_new_value[initializer] = new_initializer
+            logger.debug("Adding initializer %s", new_initializer.name)
+
+        # 4. Create nodes in the new graph from the old graph.
+        for old_node in old_graph:
+            # 4.0. Iterate and update node inputs.
+            old_node_inputs: list[ir.Value] = []
+            for old_input in old_node.inputs:
+                assert old_input is not None
+                if old_input in old_value_to_new_value:
+                    old_node_inputs.append(old_value_to_new_value[old_input])
+                else:
+                    old_node_inputs.append(old_input)
+            # 4.1. Construct the (node, inputs, attributes) hash to
+            #      check if the node is a common subexpression.
+            # Attr is not hashable
+            attributes = {}
+            for k, v in old_node.attributes.items():
+                assert isinstance(v, ir.Attr)
+                attributes[k] = v.value
+            hash_value = hash(
+                (
+                    old_node.op_identifier(),
+                    tuple(old_node_inputs),
+                    tuple(sorted(attributes.items())),
+                )
+            )
+            # TODO(titaiwang): Subgraphs are not supported yet.
+            # TODO(titaiwang): Skip control flow nodes?
+            # 4.2. Check if the node is a common subexpression.
+            if hash_value in old_node_hash_to_new_node:
+                # 4.2.1. If it is, this node is already in the new graph, so
+                #         we don't need to create a new node.
+                modified = True
```

Comment:
To convert this pass to an in-place one: we just need to modify the input graph in this one place: we just need to call `ir.convenience.replace_nodes_and_values` to delete the current node, and replace the output-values of current node with the output-values of reused-node. (If necessary, we can modify the replace_nodes_and_values utility to fit the needs ... eg., to control what information we want to retain from old and new values.)

Reply:
Done